Modules/mathmodule.c

```diff
@@ -3252,26 +3354,55 @@ perm_comb_small(unsigned long long n, unsigned long long k, int iscomb)
 
     /* For small enough n and k the result fits in the 64-bit range and can
      * be calculated without allocating intermediate PyLong objects. */
-    if (iscomb
-        ? (k < Py_ARRAY_LENGTH(fast_comb_limits)
-           && n <= fast_comb_limits[k])
-        : (k < Py_ARRAY_LENGTH(fast_perm_limits)
-            && n <= fast_perm_limits[k]))
-    {
-        unsigned long long result = n;
-        if (iscomb) {
+    if (iscomb) {
+        if (k < Py_ARRAY_LENGTH(fast_comb_limits1) && n <= fast_comb_limits1[k]) {
+            /*
+                For 0 <= k <= n <= 67, comb(n, k) always fits into a uint64_t.
+                We compute it as
+
+                    comb_odd_part << shift
+
+                where 2**shift is the largest power of two dividing comb(n, k)
+                and comb_odd_part is comb(n, k) >> shift. comb_odd_part can be
+                calculated efficiently via arithmetic modulo 2**64, using three
+                lookups and two uint64_t multiplications, while the necessary
+                shift can be computed via Kummer's theorem: it's the number of
+                carries when adding k to n - k in binary, which in turn is the
+                number of set bits of n ^ k ^ (n - k).
+            */
+            uint64_t comb_odd_part = reduced_factorial_odd_part[n]
+                                   * inverted_factorial_odd_part[k]
+                                   * inverted_factorial_odd_part[n - k];
+            int shift = _Py_popcount32((uint32_t)(n ^ k ^ (n - k)));
+            return PyLong_FromUnsignedLongLong(comb_odd_part << shift);
+        }
+
+        if (k < Py_ARRAY_LENGTH(fast_comb_limits2) && n <= fast_comb_limits2[k]) {
+            unsigned long long result = n;
             for (unsigned long long i = 1; i < k;) {
                 result *= --n;
                 result /= ++i;
             }
+            return PyLong_FromUnsignedLongLong(result);
         }
-        else {
+    }
+    else {
+        if (k < Py_ARRAY_LENGTH(fast_perm_limits) && n <= fast_perm_limits[k]) {
+            if (n <= 127) {
+                uint64_t perm_odd_part = reduced_factorial_odd_part[n]
+                                    * inverted_factorial_odd_part[n - k];
+                int shift = k - _Py_popcount32((uint32_t)n)
```

Comment:
Looks like some casts are needed here to silence the compiler on Windows

Reply:
Will fix this after merging your PR. The code for bits count will be rewritten in any case.